### PR TITLE
rewrites attributes for BlockTime to be safer with type checking

### PR DESCRIPTION
**Pull request by dpaiton**

### CODE REVIEW

File: elfpy/time/time.py [hunk 1]
```diff
@@ -154,34 +154,97 @@ def time_to_days_remaining(time_remaining: float, time_stretch: float = 1, norma
...
+        """Get the time"""
+        return self._time
+
+    @time.setter
+    def time(self, value):
+        """Private time attribute and time property cannot be set directly,
+        since we want to check types using a setter."""
+        raise AttributeError("time is a read-only attribute; use `set_time()` to adjust")
 
     def set_time(self, time: FixedPoint) -> None:
```
Comment (wakamex) [inline elfpy/time/time.py:209]: could have another function that takes in a `TimeUnit`, converts to years then sets the time in years. that is:
```py
def set_time(self, time: FixedPoint, time_unit: TimeUnit) -> None:
	self.set_time(_convert_time(from=time_unit, to="years"))
```

File: elfpy/time/time.py [hunk 2]
```diff
@@ -154,34 +154,97 @@ def time_to_days_remaining(time_remaining: float, time_stretch: float = 1, norma
...
         """ticks the time by step_size"""
-        self.time += self.step_size
+        self._time += self.step_size
+
+    def time_conversion(self, unit: TimeUnit = "seconds") -> FixedPoint:
+        """Convert time to different units
+
+        .. todo:: we will need to add conditions for self.time_unit in each conversion type
+        """
+        if unit == "seconds":  # 31,556,952 seconds in a year
```
Comment (wakamex) [inline elfpy/time/time.py:186]: storing TimeUnit as an Enum with the relevant scalar as the number would eliminate this if else block

File: tests_fp/test_time_fp.py [hunk 3]
```diff
@@ -116,17 +116,67 @@ def test_block_time_init(self):
...
+        with self.assertRaises(ValueError):
+            test_time = time.BlockTimeFP(time_unit="seconds")
 
     def test_block_time_setters(self):
         """Test attribute setters for block time"""
         test_time = time.BlockTimeFP()
-        # default step_size
+        # time
+        assert test_time.time == FixedPoint(0)
+        test_time.set_time(FixedPoint("4.0"))
```
Comment (wakamex) [inline tests_fp/test_time_fp.py:127]: letting set_time accept `str` input and doing this cast to `FixedPoint` inside `set_time` would avoid a lot of boilerplate

**Review by wakamex — CHANGES_REQUESTED** [review]

File: elfpy/time/time.py [hunk 4]
```diff
@@ -154,34 +154,97 @@ def time_to_days_remaining(time_remaining: float, time_stretch: float = 1, norma
...
-    r"""Global time."""
+    r"""State class for tracking block timestamps and global time"""
 
-    time: FixedPoint = FixedPoint(0)  # time in years
-    block_number: FixedPoint = FixedPoint(0)
-    step_size: FixedPoint = FixedPoint("1.0") / FixedPoint("365.0")  # defaults to 1 day
+    _time: FixedPoint = FixedPoint(0)
+    _block_number: FixedPoint = FixedPoint(0)
+    _step_size: FixedPoint = FixedPoint("1.0") / FixedPoint("365.0")  # defaults to 1 day
+    time_unit: str = "years"
```
[the remark this answers is not in the document]
Comment (dpaiton) [reply elfpy/time/time.py:167]: Good catch!

File: elfpy/time/time.py [hunk 5]
```diff
@@ -154,34 +154,97 @@ def time_to_days_remaining(time_remaining: float, time_stretch: float = 1, norma
...
 
-    @property
-    def time_in_seconds(self) -> FixedPoint:
-        """1 year = 31,556,952 seconds"""
-        return self.time * FixedPoint("31_556_952.0")
+    def __post_init__(self):
+        if self.time_unit != "years":
+            raise ValueError(f"Only `years` is supported for {self.time_unit}")
```
[the remark this answers is not in the document]
Comment (dpaiton) [reply elfpy/time/time.py:171]: Yeah I thought about that, but I ultimately decided against it. I didn't want to add an untested feature, and I didn't want to write tests for a feature that won't stick around. The "time refactor" is coming soon, so this is just a holdover scaffolding until that arrives.

File: elfpy/time/time.py (hunk 1 above)
Comment (dpaiton) [reply elfpy/time/time.py:209]: the conversion function only goes from years to `time_unit`. It's not any unit -to- any unit. And as above, I didn't want to write all of that out until we decide how we want the time refactor to look.

File: elfpy/time/time.py (hunk 2 above)
Comment (dpaiton) [reply elfpy/time/time.py:186]: Ok, I did this. But really though we will want to redo this with stdlib time utilities soon. I just wanted to fix the errors with the setters.

File: elfpy/time/time.py [hunk 6]
```diff
@@ -154,34 +154,109 @@ def time_to_days_remaining(time_remaining: float, time_stretch: float = 1, norma
     return normed_days_remaining * normalizing_constant
 
 
+class TimeUnit(Enum):
+    r"""Time units, with their parts in a year
+
+    FP values are stored as parts in a year as FixedPoint 18 decimal integers"""
+    SECONDS = FixedPoint(31_556_952 * 10**18)
```
Comment (wakamex) [inline elfpy/time/time.py:161]: such densely packed information 🤤

**Review by wakamex — APPROVED** [review]
very nice 96% patch coverage!

**Pull request merged**